crates/runtime/src/model/eval.rs

```diff
@@ -0,0 +1,445 @@
+                Some(
+                    arr_str
+                        .iter()
+                        .map(Option::unwrap_or_default)
+                        .collect::<Vec<&str>>(),
+                )
+            } else {
+                arr.as_any()
+                    .downcast_ref::<StringViewArray>()
+                    .map(|arr_str| {
+                        arr_str
+                            .iter()
+                            .map(Option::unwrap_or_default)
+                            .collect::<Vec<&str>>()
+                    })
+            }
+        };
+        if let Some(from_str) = from_str_opt {
+            return Ok(from_str.into_iter().map(Self::from_raw).collect());
+        }
+
+        // Try as [`DatasetOutput::Choices`].
+        let list_arr =
+            arr.as_any()
+                .downcast_ref::<ListArray>()
+                .ok_or_else(|| Error::InvalidOutputFormat {
+                    reason: "must be a string or list, but was neither".to_string(),
+                })?;
+
+        let mut result = Vec::with_capacity(list_arr.len());
+        for i in 0..list_arr.len() {
+            if list_arr.is_null(i) {
+                return Err(Error::InvalidOutputFormat {
+                    reason: "elements cannot be null".to_string(),
+                });
+            }
+            let arr = list_arr.value(i);
+            let struct_arr = arr.as_any().downcast_ref::<StructArray>().ok_or_else(|| {
+                Error::InvalidOutputFormat {
+                    reason: "must be a string or list, but was neither".to_string(),
+                }
+            })?;
+
+            let json_value = rb_to_json_value(&RecordBatch::from(struct_arr)).map_err(|e| {
+                Error::InvalidOutputFormat {
+                    reason: format!("could not convert element into JSON representation: {e}"),
+                }
+            })?;
+
+            match Self::try_from_value(json_value) {
+                Ok(Some(v)) => result.push(v),
+                Ok(None) => {
+                    return Err(Error::InvalidOutputFormat { reason: "could not convert valid list-type elements into a known model output format".to_string() });
+                }
+                Err(e) => {
+                    return Err(Error::InvalidOutputFormat { reason: format!("could not convert JSON format of an element into a known model output format: {e}") });
+                }
+            };
+        }
+
+        Ok(result)
+    }
+}
+
+#[allow(clippy::implicit_hasher)]
+pub async fn run_eval(
+    eval: &Eval,
+    df: Arc<DataFusion>,
+    model: &dyn Chat,
+    scorers: &HashMap<String, Arc<dyn Scorer>>,
+) -> Result<HashMap<String, Vec<(String, f32)>>> {
+    let Eval {
+        name: eval_name,
+        scorers: scorer_names,
+        dataset: dataset_str,
+        ..
+    } = eval;
+
+    let mut scorers_subset = HashMap::with_capacity(scorer_names.len());
```

Comment:
Just curious about the variable naming here, why is it called scorers_subset, it seems to contain all the scores used by eval

Reply:
Not amazing naming, `scorers` is all available scorers (spiced-wide). `scorers_subset` is the subset of all scorers needed for eval.

Reply:
This naming will go away in the followup PR #3780